integrationtests/src/main/java/io/kroxylicious/proxy/testcluster/ContainerBasedKafkaCluster.java

```diff
@@ -0,0 +1,235 @@
+        return brokers.stream()
+                .map(b -> String.format("localhost:%d", b.getMappedPort(KAFKA_PORT)))
+                .collect(Collectors.joining(","));
+    }
+
+    private Stream<GenericContainer<?>> allContainers() {
+        return Stream.concat(
+                this.brokers.stream(),
+                Stream.ofNullable(this.zookeeper));
+    }
+
+    @Override
+    @SneakyThrows
+    public void start() {
+        try {
+            if (zookeeper != null) {
+                zookeeper.start();
+                Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                    Container.ExecResult result = this.zookeeper.execInContainer(
+                            "sh", "-c",
+                            String.format("_JAVA_OPTIONS=-Dlogback.configurationFile=/tmp/zookeeper_logback.xml zkCli.sh -server zookeeper:%s whoami", ZOOKEEPER_PORT));
+                    LOGGER.debug("zookeeper running {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                    return result.getExitCode() == 0;
+                });
+            }
+            Startables.deepStart(brokers.stream()).get(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException | ExecutionException | TimeoutException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            stop();
+            throw new RuntimeException("startup failed or timed out", e);
+        }
+
+        if (clusterConfig.isKraftMode()) {
+            Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                Container.ExecResult result = this.brokers.iterator().next().execInContainer(
+                        "sh", "-c",
+                        "kafka-metadata-shell.sh --snapshot /bitnami/kafka/data/__cluster_metadata-0/*.log ls /brokers");
+                LOGGER.debug("kraft cluster ready probe {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
+                Optional<String> brokers = Optional.ofNullable(result.getStdout());
+                return brokers.map(b -> Arrays.stream(b.split(System.lineSeparator())).count()).map(count -> count.intValue() == this.clusterConfig.getBrokersNum())
+                        .orElse(false);
+            });
+        }
+        else {
+            Unreliables.retryUntilTrue(READY_TIMEOUT_SECONDS, TimeUnit.SECONDS, () -> {
+                Container.ExecResult result = this.zookeeper.execInContainer(
+                        "sh", "-c",
+                        String.format("_JAVA_OPTIONS=-Dlogback.configurationFile=/tmp/zookeeper_logback.xml zkCli.sh -server zookeeper:%s ls /brokers/ids | tail -n 1",
+                                ZOOKEEPER_PORT));
+                LOGGER.debug("zookeeper cluster ready probe {} / {} / exit {}", result.getStdout(), result.getStderr(), result.getExitCode());
```

Comment:
It is a little confusing to say this is a `zookeeper cluster` without mentioning kafka. Maybe `Kafka cluster with zookeeper`?

Reply:
I pushed the responsibility to wait for zookeeper and kafka cluster to be ready to the containers.